### DISCUSSION TITLE
Server Action - how to store file in tmp directory

### DISCUSSION THREAD

**Discussion by andrewmcoupe** [vercel/next.js#57609] [Help]
### Summary

I have a server action that receives a file using FormData. I'd like to upload the file to Cloudinary - the Cloudinary API expects me to pass a string which is the path to the file. 

Because the server action is running on the server it has no concept of a file path so I'm thinking that I need to store the file locally i.e. in a temp directory and then I can pass this path to the Cloudinary API.

How can I store this file in a temporary directory?

**Comment by icyJoseph** [comment]
Mmm something like:

```tsx
import { saveFile } from "./_action";

export default function Home() {
  return (
    <form action={saveFile}>
      <input type="file" name="file" />
      <button type="submit">Go</button>
    </form>
  );
}
```

Where the server action is:

```ts
"use server";
import fs from "fs/promises";

export async function saveFile(formData: FormData) {
  const file = formData.get("file") as File;

  const data = await file.arrayBuffer();

  await fs.writeFile(`/tmp/${file.name}`, Buffer.from(data));
}
```

Vercel allows `/tmp` access IIRC: https://vercel.com/guides/how-can-i-use-files-in-serverless-functions#using-temporary-storage

**Reply by andrewmcoupe** [reply]
I did try this but the server action couldn't read from /tmp

**Reply by icyJoseph** [reply]
Could you share some code/screenshot?

**Reply by icyJoseph** [reply]
And can't the API you are using, ever take a Buffer or something else?

**Reply by andrewmcoupe** [reply]
Cloudinary can take a buffer apparently - this works locally but on a next production build their server returns a 500 for some reason. They don't seem to want to help either so I'm just going to upload to S3 instead.

**Reply by mathealves** [reply]
This page doesn't contain any info on temporary storage anymore, and I can't find that info anywhere. Was it discontinued or something?

**Comment by mathealves** [comment]
Having this issue at the moment. Tried   `const filePath = path.join(process.cwd(), "/tmp/input.wav");` but get "Error processing audio: Error: ENOENT: no such file or directory, open '/var/task/tmp/input.wav'" from Vercel

**Reply by Adilovicc** [reply]
Same here, did you find a solution?

**Comment by ogbotemi-2000** [comment]
Hello, I recently migrated my web app to Vercel from Railway and developed a working solution, which can be found here: *https://rmrf-css.vercel.app*, for the topic of this discussion.
Here are my observations and solutions

1. On Railway, everything works as expected of a Node.js server and the global object in my `server.js` file stores data throughout the runtime of the server.
2. Ensuring this behavior on _serverless_ Vercel required mimicking storing and reading values in objects by using the file system as shown below and reading/writing to the `/tmp/` directory to avoid errors due to a violation of writing files on the server.
3. To avoid the issue of having to serialize URLs into valid filenames for saving their responses, a uuid is generated and used as the filename of the URLs' responses, it is then sent to the client among the response for the initial request and sent back by the client to be used as a unique string bearing the name of the file to be read similar to using an object in uniqueness and versatility.

```js
/* store.js */
let fs     = require('fs'),
    uuid   = (a) => (a ? (a ^ ((Math.random() * 16) >> (a / 4))).toString(16) : ([1e7] + -1e3 + -4e3 + -8e3 + -1e11).replace(/[018]/g, uuid)),
    /* may be made asynchronous */
    rm     = id=>fs.unlinkSync(path.join(dir, id+'.txt')),
    dir    = '/tmp/',
    path   = require('path');

/*writes/reads data as text/plain, open to modifications*/
module.exports = {
  read: id =>new Promise((resolve, reject)=>fs.readFile(dir+id+'.txt', (err, buffer)=>{
      if(err) reject(err);
      else resolve(buffer.toString())
    })
  ),
  write: function(data, cb, id) {
    fs.writeFile(dir+(id=uuid())+'.txt', data, _=>cb(id))
  },
  rm 
}
```

Here is how I use it and how it may be used in the wild

```js
let store  = require('./store');
module.exports = async function(request, response, uuid) {
  let {url, id} = request.query;
  if(uuid = await store.read(id)) {
    /* exists */
    response.send(/* data */),
    /* you may remove the file or leave it since the contents of the /tmp/ directory are likely temporary */
    store.rm(uuid)
  } else {
    store.write('data', function(id) {
       /* send the unique id for the stored data in some way as JSON or text */
       response.send(id)
    })
  }
}
```

I hope it serves as a viable workaround to the issue with using global variables in Vercel serverless backends
